Document .http file syntax and VS .http file editor (#29218)
dotnet/AspNetCore.Docs: aspnetcore/test/http-files.md

## Body

Add the request body after a blank line, as shown in the following example:

```
POST https://localhost:7220/weatherforecast
Content-Type: application/json
Accept-Language: en-US,en;q=0.5

{
    "date": "2023-05-10",
    "temperatureC": 30,
    "summary": "Warm"
}

###
```

## Unsupported syntax

The Visual Studio 2022 `.http` file editor was inspired by the [Visual Studio Code REST Client extension](https://marketplace.visualstudio.com/items?itemName=humao.rest-client) but does not have all the features of the Visual Studio Code extension. Some of the more significant features that the Visual Studio Code extension has but Visual Studio 2022 doesn't have are the following:

* Optional HTTP method
* Request line that spans more than one line
* Named requests
* Dynamic variables
* Environment files
* Specify file path as body of the request
* Mixed format for body when using multipart/form-data
* GraphQL requests
* cURL request
* Copy/paste as cURL
* Request history
* Save response body to file
* Certificate based authentication
* Prompt variables
* System variables
* Customize response preview
* Per-request settings

> The Visual Studio 2022 `.http` file editor is sill under development, and some of these features might be added in the future. For more information about the Visual Studio Code extension, see [REST Client extension](https://marketplace.visualstudio.com/items?itemName=humao.rest-client).

Reviewer: ```suggestion
The Visual Studio 2022 `.http` file editor is still under development, and some of these features might be added in the future. For more information about the Visual Studio Code extension, see [REST Client extension](https://marketplace.visualstudio.com/items?itemName=humao.rest-client).
```

Author: Thanks, @gfoidl !

## Prerequisites

* [Visual Studio 2022 version 17.6 Preview 6 or later](https://visualstudio.microsoft.com/vs/preview/) with the **ASP.NET and web development** workload installed.

## `.http` file syntax

The following sections explain `.http` file syntax.

### Comments

Lines that start with either `#` or `//` are comments. These lines are ignored when Visual Studio sends HTTP requests.

### Variables

A line that starts with `@` defines a variable by using the syntax `@VariableName=Value`.

Variables can be referenced in requests that are defined later in the file. They are referenced by wrapping their names in double curly braces (`{{` and `}}`). The following example shows two variables defined and then used in a request:

```
@hostname=localhost
@port=44320
GET https://{{hostname}}:{{port}}/weatherforecast
```

Variables can be defined using values of other variables that were defined earlier in the file. The following example uses one variable in the request instead of the two shown in the preceding example:

```
@hostname=localhost
@port=44320

> @host={{hostname}}:{{port}}
> GET https://{{host}}/api/search/tool

Reviewer 2: ```suggestion
@hostPort={{hostname}}:{{port}}
GET https://{{hostPort}}/api/search/tool
```

```

### Requests

The format for an HTTP request is `HTTPMethod URL HTTPVersion`, all on one line, where:

* `HTTPMethod` is the HTTP method to use, for example:

>   * OPTIONS
>   * GET
>   * HEAD
>   * POST
>   * PUT
>   * PATCH
>   * DELETE
>   * TRACE
>   * CONNECT

Reviewer 2: ```suggestion
  * [OPTIONS](https://developer.mozilla.org/docs/Web/HTTP/Methods/OPTIONS)
  * [GET](https://developer.mozilla.org/docs/Web/HTTP/Methods/GET)
  * [HEAD](https://developer.mozilla.org/docs/Web/HTTP/Methods/HEAD)
  * [POST](https://developer.mozilla.org/docs/Web/HTTP/Methods/POST)
  * [PUT](https://developer.mozilla.org/docs/Web/HTTP/Methods/put)
  * [PATCH](https://developer.mozilla.org/docs/Web/HTTP/Methods/PATCH)
  * [DELETE](https://developer.mozilla.org/docs/Web/HTTP/Methods/DELETE)
  * [TRACE](https://developer.mozilla.org/docs/Web/HTTP/Methods/TRACE)
  * [CONNECT](https://developer.mozilla.org/docs/Web/HTTP/Methods/CONNECT)
```

* `URL` is the URL to send the request to. The URL can include query string parameters. The URL doesn't have to point to a local web project. It can point to any URL that Visual Studio can access.
* `HTTPVersion` is optional and specifies the HTTP version that should be used, that is, `HTTP/1.1`, `HTTP/2`, or `HTTP/3`.

> A file can contain multiple requests by using lines with `###` as delimiters. The following example showing 3 requests in a file illustrates this syntax:

Reviewer 2: ```suggestion
A file can contain multiple requests by using a line with `###` as a delimiter. The following example showing 3 requests in a file illustrates this syntax:
```

```
GET https://localhost:7220/weatherforecast

###

GET https://localhost:7220/weatherforecast?date=2023-05-11&location=98006

###

GET https://localhost:7220/weatherforecast HTTP/3

###
```

## Request headers

> To add one or more headers, add each header on its own line immediately after the request line (no blank lines). The format is `HeaderName: Value`, as shown in the following examples:

Reviewer 2: ```suggestion
To add one or more headers, add each header on its own line immediately after the request line. Don't add a blank lines. The format is `HeaderName: Value`, as shown in the following examples:
```

```
GET https://localhost:7220/weatherforecast
Date: Wed, 27 Apr 2023 07:28:00 GMT

###

GET https://localhost:7220/weatherforecast
Cache-Control: max-age=604800
Age: 100

###
```

> [!IMPORTANT]

> > When calling an API that authenticates with headers, be careful to not commit any secrets to a source code repository. The Visual Studio 2022 `.http` file editor doesn't have any special handling for secrets.

Reviewer 2: ```suggestion
> When calling an API that authenticates with headers, do **_not_** commit any secrets to a source code repository. The Visual Studio 2022 `.http` file editor doesn't secure secrets.
```

### Body

Add the request body after a blank line, as shown in the following example:

```
POST https://localhost:7220/weatherforecast
Content-Type: application/json
Accept-Language: en-US,en;q=0.5

{
    "date": "2023-05-10",
    "temperatureC": 30,
    "summary": "Warm"
}

###
```

### Unsupported syntax

> The Visual Studio 2022 `.http` file editor doesn't have all the features that the Visual Studio Code [REST Client extension](https://marketplace.visualstudio.com/items?itemName=humao.rest-client) has. The following list includes some of the more significant features that the Visual Studio Code extension has but Visual Studio 2022 doesn't have:

Reviewer 2: ```suggestion
The Visual Studio 2022 `.http` file editor doesn't have all the features that the Visual Studio Code [REST Client extension](https://marketplace.visualstudio.com/items?itemName=humao.rest-client) has. The following list includes some of the more significant features available only in the Visual Studio Code extension:
```

Less is more.

## Use Endpoints Explorer

**Endpoints Explorer** is a tool window in Visual Studio 2022 that provides a UI that integrates with the `.http` file editor for testing HTTP requests.

### Enable Endpoints Explorer

**Endpoints Explorer** is still a preview feature, so you have to enable it in Visual Studio 2022. To do that, select **Tools** > **Options** > **Environment** > **Preview Features** > **Web API Endpoints Explorer**.

### Open Endpoints Explorer

Select **View** > **Other Windows** > **Endpoints Explorer**.

### Add a request to an `.http` file

Right-click a request in **Endpoints Explorer** and select **Generate Request**.

:::image type="content" source="~/test/http-files/_static/generate-request.png" alt-text="Endpoints Explorer window showing request context menu with 'Generate Request' menu selection highlighted.":::

* If an `.http` file is open in the editor, the request is added to that file.
* If an `.http` file isn't open and one exists in the project, the file is opened and the request is added to that file.
* If no `.http` file exists in the project, one is created with the project name as the file name, and the request is added to that file.

> The preceding screenshot shows endpoints defined by the minimal API project template, and the following example shows the request that is generated for the selected endpoint:

Reviewer 2: ```suggestion
The preceding screenshot shows endpoints defined by the minimal API project template. The following example shows the request that is generated for the selected endpoint:
```
